grr/server/grr_response_server/gui/ui/components/client_details/client_diff.ts

```diff
@@ -202,3 +202,59 @@ export function getClientVersions(clientSnapshots: Client[]):
 
   return clientChanges;
 }
+
+function getFirstStringsJoinedPath(path: any[]): string {
+  let tokens: string[] = [];
+  path.forEach(val => {
+    if (typeof val !== 'string') return;
+    tokens.push(val);
+  });
+
+  return tokens.join('.');
+}
+
+function getPathsOfChangedEntries(differences: Diff<Client, Client>[]):
+    string[] {
+  const pathsSetOfChangedEntries: Set<string> = new Set();
+
+  differences.forEach(diffItem => {
+    if (diffItem.path === undefined) return;
+    const joinedPath = getFirstStringsJoinedPath(diffItem.path);
+    if (!RELEVANT_ENTRIES_LABEL_MAP.has(joinedPath)) return;
+
+    pathsSetOfChangedEntries.add(joinedPath);
+  });
+
+  return [...pathsSetOfChangedEntries.values()];
+}
+
+/**
+ * Returns relevant entry paths mapped to the client snapshots in which the
+ * changes were introduced, reverse chronologically ordered. A property path may
+ * not be inside the map if no changes were applied to it since it's creation
+ * @param clientSnapshots an array of chronologically reverse ordered client
+ *     snapshots
+ */
+export function getClientEntriesChanged(clientSnapshots: Client[]):
+    Map<string, ReadonlyArray<Client>> {
+  const clientChangedEntries: Map<string, Client[]> = new Map();
+
+  for (let i = 0; i < clientSnapshots.length - 1; i++) {
+    const differences = diff(clientSnapshots[i + 1], clientSnapshots[i]);
```

Comment:
This accesses `clientSnapshots[clientSnapshots.length]` and will return `undefined`. Is this intended? In general, this code mixes low-level iteration logic and high-level diffing logic. I suggest extracting the logic for pairwise iteration into a generic helper function, e.g. `function pairwise<T>(arr: T[]) => ReadonlyArray<[T, T]>`.

Also, use functional programming like `filter(difference => difference !== undefined)` over branching and side-effects.

Reply:
The `i` variable will take values in the range `[0, clientSnapshots.length - 2]`, because of the condition inside the for statement: `i < clientSnapshots.length - 1`. This means that `i + 1` will be in the range `[1, clientSnapshots.length - 1]`. Knowing this, I don't understand how it accesses `clientSnapshots[clientSnapshots.length]`. On the other hand, the `pairwise` helper function is a good idea, I will make the changes.